File: src/services/whois.service.ts
/**
 * Whois Service - Perform whois lookups
 */
import { exec } from 'child_process';
import { promisify } from 'util';

const execAsync = promisify(exec);

export class WhoisService {
  async getWhois(domainOrIP: string): Promise<any> {
    try {
      const { stdout } = await execAsync(`whois ${domainOrIP}`, { timeout: 10000 });
      
      // Parse whois output (basic parsing)
      const lines = stdout.split('\n');
      const result: any = {
        domain: domainOrIP,
        raw: stdout
      };

      for (const line of lines) {
        if (line.includes('Registrar:')) {
          result.registrar = line.split('Registrar:')[1]?.trim();
        }
        if (line.includes('Creation Date:') || line.includes('Created:')) {
          result.creationDate = line.split(':')[1]?.trim();
        }
        if (line.includes('Expiration Date:') || line.includes('Expires:')) {
          result.expirationDate = line.split(':')[1]?.trim();
        }
        if (line.includes('Name Server:') || line.includes('Name servers:')) {
          if (!result.nameServers) result.nameServers = [];
          result.nameServers.push(line.split(':')[1]?.trim());
        }
      }

      return result;
    } catch (error: any) {
      console.error('Error getting whois:', error);
      return null;
    }
  }
}

File: services/whois_service.py
"""
Whois Service - Perform whois lookups
"""
import whois
import asyncio
from typing import Optional, Dict, Any
from utils.helpers import is_valid_ip


class WhoisService:
    """Service for performing whois lookups"""
    
    async def get_whois(self, domain_or_ip: str) -> Optional[Dict[str, Any]]:
        """Get whois information for domain or IP"""
        try:
            loop = asyncio.get_event_loop()
            w = await loop.run_in_executor(
                None,
                lambda: whois.whois(domain_or_ip)
            )
            
            if w:
                return {
                    "domain": w.domain,
                    "registrar": w.registrar,
                    "creation_date": str(w.creation_date) if w.creation_date else None,
                    "expiration_date": str(w.expiration_date) if w.expiration_date else None,
                    "updated_date": str(w.updated_date) if w.updated_date else None,
                    "name_servers": w.name_servers if w.name_servers else [],
                    "status": w.status if w.status else [],
                    "emails": w.emails if w.emails else [],
                    "org": w.org,
                    "country": w.country,
                }
        except Exception as e:
            print(f"Error getting whois: {e}")
            return None
        
        return None

